### PR TITLE
#2663 Adding KafkaStreams quickstart example

### PR DESCRIPTION
**Pull request by gunnarmorling**
Hey @gsmet, @cescoffier, here's the quickstart example for the Kafka Streams extension. Feedback welcome!

(Requires https://github.com/quarkusio/quarkus/pull/2794)

### CODE REVIEW

File: kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/KafkaStreamsPipeline.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    private static final String WEATHER_STATIONS_STORE = "weather-stations-store";
+
+    private static final String WEATHER_STATIONS_TOPIC = "weather-stations";
+    private static final String TEMPERATURE_VALUES_TOPIC = "temperature-values";
+    private static final String TEMPERATURES_AGGREGATED_TOPIC = "temperatures-aggregated";
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaStreamsPipeline.class);
+
+    @ConfigProperty(name="org.acme.quarkus.sample.kafkastreams.bootstrap.servers", defaultValue="localhost:9092")
+    private String bootstrapServers;
```
Comment (jpechane) [inline kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/KafkaStreamsPipeline.java]: @gunnarmorling Use correct access modifiers
```
[INFO] [io.quarkus.arc.processor.BeanProcessor] Found unrecommended usage of private members (use package-private instead) in application beans:
	- @Inject field org.acme.quarkus.sample.kafkastreams.KafkaStreamsPipeline#bootstrapServers,
	- @Inject field org.acme.quarkus.sample.kafkastreams.KafkaStreamsPipeline#host,
	- @Inject field org.acme.quarkus.sample.kafkastreams.KafkaStreamsPipeline#port
	- and 1 more - please enable debug logging to see the full list
```

File: kafka-streams-quickstart/README.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+For development purposes it can be handy to run the _producer_ and _aggregator_ applications
+directly on your local machine instead of via Docker.
+For that purpose, a separate Docker Compose file is provided which just starts Apache Kafka and ZooKeeper,
+configured to be accessible from your host system:
+
+```bash
+docker-compose -f docker-compose-local.yaml up
+
+mvn compile quarkus:dev -f producer/pom.xml
+mvn compile quarkus:dev -Dquarkus.http.port=8081 \
```
Comment (jpechane) [inline kafka-streams-quickstart/README.md]: The command fails when `HOSTNAME` is not set

Comment (gunnarmorling) [reply kafka-streams-quickstart/README.md]: Added.

File: kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/KafkaStreamsPipeline.java (hunk 1 above)
Comment (gunnarmorling) [reply kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/KafkaStreamsPipeline.java]: Done.

**Comment by gunnarmorling** [comment]
Reworked as per @jpechane's comments.

**Comment by gunnarmorling** [comment]
@gsmet, @cescoffier, any feedback from your side? I'll be working on the quickstart page next, so if you see the need for changes to the code itself, please let me know. Thanks!

**Comment by gunnarmorling** [comment]
@cescoffier, what's the tool you used for creating that image in the Kafka quickstart guide? If possible, I'd use the same for the Kafka Streams guide. Thanks!

**Comment by cescoffier** [comment]
It's done using a combination of hand-made pictures and Keynote. Send me a
draft version and I would produce the picture.

Le dim. 16 juin 2019 à 11:02, Gunnar Morling <notifications@github.com> a
écrit :

File: kafka-streams-quickstart/.gitignore [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+# patch
+*.orig
+*.rej
+
+# Maven
+target/
+pom.xml.tag
+pom.xml.releaseBackup
+pom.xml.versionsBackup
+release.properties
```
Comment (gsmet) [inline kafka-streams-quickstart/.gitignore]: Is there any reason why you added that in a subdirectory? Shouldn't we just have a global `.gitignore`?

File: kafka-streams-quickstart/aggregator/pom.xml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.rest-assured</groupId>
+      <artifactId>rest-assured</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-kafka-streams</artifactId>
```
Comment (gsmet) [inline kafka-streams-quickstart/aggregator/pom.xml]: Maybe reorganize to have this one at the top?

File: kafka-streams-quickstart/producer/pom.xml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.rest-assured</groupId>
+      <artifactId>rest-assured</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-smallrye-reactive-messaging-kafka</artifactId>
```
Comment (gsmet) [inline kafka-streams-quickstart/producer/pom.xml]: Same here: I would have that one on top.

File: kafka-streams-quickstart/.gitignore (hunk 3 above)
Comment (gunnarmorling) [reply kafka-streams-quickstart/.gitignore]: It was generated by Quarkus :) I'll remove it.

Comment (gunnarmorling) [reply kafka-streams-quickstart/.gitignore]: It was generated by Quarkus :) I'll remove it.

Comment (gunnarmorling) [reply kafka-streams-quickstart/.gitignore]: It was generated by Quarkus :) I'll remove it.

File: kafka-streams-quickstart/aggregator/pom.xml (hunk 4 above)
Comment (gunnarmorling) [reply kafka-streams-quickstart/aggregator/pom.xml]: Good idea. Done.

File: kafka-streams-quickstart/producer/pom.xml (hunk 5 above)
Comment (gunnarmorling) [reply kafka-streams-quickstart/producer/pom.xml]: Done.

**Comment by gunnarmorling** [comment]
Pushed a commit to address @gsmet's remarks.

File: kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/KafkaStreamsPipeline.java [hunk 6]
```diff
@@ -171,25 +178,19 @@ private void waitForTopicsToBeCreated(String bootstrapServers) {
...
-                try {
-                    Thread.sleep(1_000);
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
+                if (topicNames.contains(WEATHER_STATIONS_TOPIC) && topicNames.contains(TEMPERATURE_VALUES_TOPIC)) {
+                    return;
                 }
+
+                LOG.info("Waiting for topics {} and {} to be created", WEATHER_STATIONS_TOPIC, TEMPERATURE_VALUES_TOPIC);
+                Thread.sleep(1_000);
```
Comment (gsmet) [inline kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/KafkaStreamsPipeline.java]: @cescoffier mentioned Awaitility somewhere. Is there a reason why you don't use it? (I might have missed something)

Comment (gunnarmorling) [reply kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/KafkaStreamsPipeline.java]: So far I've considered Awaitility more as a testing utility (which is why it made sense to me to use it in the integration test). But is it something you'd indeed use in prod code (which more is what this example is supposed to show)? Also it'd mean adding another dependency to the quickstart. But I'm open.

Comment (cescoffier) [reply kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/KafkaStreamsPipeline.java]: I don't have a strong opinion here. Awaitility won't help much in this case and it is a pattern we find a lot when using kafka. So I'm good with the current version.

File: kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/JsonbSerde.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import javax.json.bind.JsonbBuilder;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serializer;
+
+/**
+ * A {@link Serde} that (de-)serializes JSON.
+ */
+public class JsonbSerde<T> implements Serde<T> {
```
Comment (cescoffier) [inline kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/JsonbSerde.java]: I would consider this as a good addition to the Kafka client extension. WDYT?

**Comment by cescoffier** [comment]
The MP Reactive Messaging config would need to be updated with the new _official_ format.

File: kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/JsonbSerde.java (hunk 7 above)
Comment (gunnarmorling) [reply kafka-streams-quickstart/aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/streams/JsonbSerde.java]: Yes, good idea. I already have filed https://github.com/quarkusio/quarkus/issues/2863 as a follow-up for the Kafka Streams extension. Just added this as one item there.

**Comment by gunnarmorling** [comment]
> The MP Reactive Messaging config would need to be updated with the new official format.

Ok, do you have a reference to the details?

**Comment by gunnarmorling** [comment]
> It would be great to have a simple test and a native test.
> We use these tests as canary to be sure the native packaging in not broken.

Don't you think the tests of the actual extension are good enough for this? There we already have a test for JVM an native. I'd largely duplicate that set-up here to this quickstart.

**Comment by cescoffier** [comment]
> > The MP Reactive Messaging config would need to be updated with the new official format.
> 
> Ok, do you have a reference to the details?

I will publish the SmallRye implementation documentation later today. But, as I know you are impatient, here is what it gives for the Kafka quickstart:

```
# Configure the Kafka sink (we write to it)
mp.messaging.outgoing.generated-price.connector=smallrye-kafka
mp.messaging.outgoing.generated-price.topic=prices
mp.messaging.outgoing.generated-price.value.serializer=org.apache.kafka.common.serialization.IntegerSerializer

# Configure the Kafka source (we read from it)
mp.messaging.incoming.prices.connector=smallrye-kafka
mp.messaging.incoming.prices.topic=prices
mp.messaging.incoming.prices.value.deserializer=org.apache.kafka.common.serialization.IntegerDeserializer
```

**Pull request merged**